include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+   *
+   * NonlinearSolver supports any vector and matrix type having constructors and
+   * methods:
+   *
+   * @code
+   * class VectorType : public Subscriptor
+   *    ...
+   *    explicit VectorType(Vec);
+   *    ...
+   *    Vec & petsc_vector();
+   *    ...
+   * @endcode
+   *
+   * @code
+   * class MatrixType : public Subscriptor
+   *    ...
+   *    explicit MatrixType(Mat);
+   *    ...
+   *    Mat & petsc_matrix();
+   *    ...
+   * @endcode
+   *
+   * In particular, the supported types are the ones that can *wrap*
+   * PETSc's native vector and matrix classes, that are able to modify
+   * them in place, and that can return PETSc native types when requested.
+   *
+   * To use the solvers the user needs to provide the implementation of $F$ via
+   * the NonlinearSolver::residual callback.
+   *
+   * The default linearization procedure of a solver instantiated with
+   * this class consists in using Jacobian-Free-Newton-Krylov; the action of
+   * tangent matrices inside a linear solver process are approximated via
+   * matrix-free finite-differencing of the nonlinear residual equations.
+   * For details, consult the [PETSc
+   * manual](https://petsc.org/release/manual/snes/#sec-nlmatrixfree).
+   *
+   * In alternative, users can also provide the implementation of the
+   * Jacobian. This can be accomplished in two ways:
+   *  - a-la-PETSc style using NonlinearSolver::jacobian
+   *  - a-la-Deal.II style using NonlinearSolver::setup_jacobian and
+   * NonlinearSolver::solve_for_jacobian_system
+   *
+   * In case both approaches are coded, the deal.II style
+   * will be used.
+   *
+   * The second approach is more in style with the deal.II philosophy
+   * and it also allows command line customization, like for example,
+   * @code
+   * ./myApp -snes_type newtontr -ksp_type cg
+   * @endcode
+   * in case the user wants to change the default nonlinear solver to
+   * a trust region solver and iterate on the tangent system with CG,
+   * still using NonlinearSolver::solve_for_jacobian_system as a preconditioner.
+   *
+   * The first approach has instead the advantage that only the matrix assembly
+   * procedure has to be coded, thus allowing quicker implementations and
+   * faster turnaround for experimenting with linear solver preconditioning
+   * configurations via command line customizations, like for example,
+   * @code
+   * ./myApp -ksp_type cg -pc_type gamg
+   * @endcode
+   *
+   * In case the nonlinear equations are derived from energy minimization
+   * arguments, it may be beneficial to perform linesearch or test trust-region
+   * model reductions using the energy functional. In such cases, users can
+   * set an optional NonlinearSolver::energy callback.
+   */
+  template <typename VectorType  = PETScWrappers::VectorBase,
+            typename PMatrixType = PETScWrappers::MatrixBase,
+            typename AMatrixType = PMatrixType>
```

Comment:
What happens if I use `LinearAlgebra::distributed::Vector<Number>` or `LinearAlgebra::distributed::BlockVector<Number>` and the don't use "a-la-deal.II style"? Does the code still compile?

Reply:
It will compile as long as the templated classes provide the implementations of the methods as per the documentation. See https://github.com/dealii/dealii/pull/15065/files#diff-bfada67eb337157b20ddaf1a0fbfc7284257f606f786c18fff7d8b08dfbc5726